### PR TITLE
Refactor portal serializer + Add email logic for updates on Content

### PR DESCRIPTION
**Pull request by vcai122**
* refactor portal serializer to use abstract class + deal with special case for form data

* set up email configs + email utils (wrapper around django email)

* set up email notifs on content save/update

### CODE REVIEW

File: backend/utils/email.py
Comment (judtinzhang) [inline backend/utils/email.py:15]: Not a huge fan of this line, especially since the name of the variable has "list" in it. Can we just enforce that `recipient_list` is always of type `list`?

File: backend/utils/email.py
Comment (judtinzhang) [inline backend/utils/email.py:14]: Any reason this is `None`?

File: backend/utils/email.py
Comment (judtinzhang) [inline backend/utils/email.py:9]: When will this be the case? I feel like if this is ever `None` we shouldn't fail silently like this (the return value of this function is never checked)

**Review by judtinzhang — COMMENTED** [review]
Looks good! Some comments and questions

File: backend/pennmobile/settings/base.py
Comment (vcai122) [reply backend/pennmobile/settings/base.py:188]: host user might be `accounts@pennlabs.org` and default from might be `Penn Mobile <pennmobile@pennlabs.org>`

File: backend/portal/models.py
Comment (vcai122) [reply backend/portal/models.py:88]: wait yes need for return here because i want need to hard stop since prev is None. 
it was originally 
`self._on_create()
return`

but i changed it to be fancy

File: backend/portal/serializers.py
Comment (vcai122) [reply backend/portal/serializers.py:67]: I think its fine, because also this means we don't have to change when we add multiselect. And even if a malicious actor sent us multiselect=True, we need to first approve it and even if we approve it won't do anything cuz the frontend code doesn't use it right now

File: backend/portal/models.py
Comment (judtinzhang) [reply backend/portal/models.py:88]: Wait then why don't you just have an `elif` for the latter clause?

File: backend/portal/serializers.py
Comment (judtinzhang) [reply backend/portal/serializers.py:67]: Sounds good

File: backend/portal/models.py
Comment (vcai122) [reply backend/portal/models.py:88]: actually yea i could. I did return because i thought i would have more conditions to send an email

File: backend/portal/serializers.py
Comment (vcai122) [reply backend/portal/serializers.py:59]: The serializer will only act on things coming in through the route, so operations directly through the shell or through admin portal does not go through the serializer. So admin comment/priority does not trigger the serializer. 

With regards to club_comment change, I agree, probably don't need to set it to draft but it seems like a bit of an edge case (can't imagine people writing comments after post is approved), and kinda don't want to deal with writing that code + it'll be an extra query to db to get the previous object.

File: backend/portal/serializers.py
Comment (vcai122) [reply backend/portal/serializers.py:244]: The reason you put target_population as read_only for Posts is because otherwise, the built in serializer tries to validate / set it, but it cannot succesfully validate the data coming from a form causing errors. (hence you marked it as read_only and then manually did it yourself after the fact).

I've tested post with image + non-empty and empty list together... after dealing with the edge case in is_valid in the Post serializer, it should be treating the target_population as any other field now

File: backend/utils/email.py
Comment (vcai122) [reply backend/utils/email.py:14]: if we set to None, it uses the default one specified in settings (base.py)

File: backend/utils/email.py
Comment (vcai122) [reply backend/utils/email.py:9]: I did this because I thought this could just be "intended bevahior" for this function for sake of ease lol.

A case where we need this right now is if a post without a user attached onto it (so old posts) are updated, the recipient list would be None. 

I guess the real question is whether this should be the intended behavior of this function. I have no preference. We can throw an error instead and do the None checking on the side of the caller of this function instead

File: backend/portal/models.py
Comment (judtinzhang) [reply backend/portal/models.py:88]: yep up to you

File: backend/portal/serializers.py
Comment (judtinzhang) [reply backend/portal/serializers.py:59]: Oh right. Okay yea sounds good for the priority and admin comment. And sure we can keep as is for now

File: backend/portal/serializers.py
Comment (judtinzhang) [reply backend/portal/serializers.py:244]: Bet thanks

File: backend/utils/email.py
Comment (judtinzhang) [reply backend/utils/email.py:9]: I think I would prefer erroring here. Definitely will save a few minutes if someone needs to debug this code later on

File: backend/portal/models.py
Comment (tuneerroy) [inline backend/portal/models.py:58]: if these are meant to be private methods, they should start with two underscores, not one.

File: backend/portal/models.py
Comment (tuneerroy) [inline backend/portal/models.py:72]: don't need to pass in the last param for `getattr` since the default return value if not found is `None`.

also... should do `if _ is not None` ... more Pythonic :)

File: backend/portal/serializers.py
Comment (tuneerroy) [inline backend/portal/serializers.py:238]: hot damn, a `list` and then a `map`...

File: backend/portal/serializers.py
Comment (tuneerroy) [inline backend/portal/serializers.py:240]: could also probably just do 
```
self.initial_data["target_population"] = [int(v) for v in self.initial_data.get("target_populations", "").split(",")]
```

**Review by tuneerroy — APPROVED** [review]
didn't test it but your test cases look comprehensive so i trust it! ow lgtm :<

File: backend/portal/models.py
Comment (vcai122) [reply backend/portal/models.py:72]: no u need the default else it errors. 

but we have this pattern everywhere in our code...

File: backend/portal/models.py
Comment (vcai122) [reply backend/portal/models.py:58]: hm
https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-single-and-double-underscore-before-an-object-name

File: backend/portal/models.py
Comment (tuneerroy) [reply backend/portal/models.py:72]: o u right

argh fine keep it

File: backend/portal/models.py
Comment (tuneerroy) [reply backend/portal/models.py:58]: damn u right, that's so weird, I've always assumed the name mangling is a good way to enforce private

File: backend/portal/models.py
Comment (minghansun1) [inline backend/portal/models.py:50]: In line 50, created_date should probably be immutable. 
created_date = models.DateTimeField(default=timezone.now, editable=False)

Update save to
def save(self, *args, **kwargs):
    if not self.pk:
        self.created_date = timezone.now()  # Set created_date to the current time
    else:
        if 'update_fields' in kwargs and 'created_date' in kwargs['update_fields']:
            kwargs['update_fields'].remove('created_date')

    prev = self.__class__.objects.filter(id=self.id).first()
    super().save(*args, **kwargs)

    if prev is None:
        self._on_create()
        return

    if self.status != prev.status:
        self._on_status_change()

File: backend/portal/serializers.py
Comment (minghansun1) [inline backend/portal/serializers.py:32]: May be more efficient to do this using set difference instead of nested loops

**Review by minghansun1 — COMMENTED** [review]
General feedback: Everything looks logical to me, but I'm not familiar with Django best practices so it's hard for me to find specific issues.

Design Consideration:
For PollVotes, we could give users the option to be anonymous or named, and let us filter results based anonymous votes, named votes, and overall votes.

Specific Issues: Make created_date immutable, do _auto_add_target_population using set difference instead of nested loops

File: backend/portal/models.py
Comment (vcai122) [reply backend/portal/models.py:50]: If you look in the serializer for content, you'll see that created_date is read only so it is enforced there. 
Technically an admin could change the created date, but there is no way to prevent this since you can always change a row in a db. In other words you can't enforce atomicity at the db level

File: backend/portal/models.py
Comment (vcai122) [reply backend/portal/models.py:58]: wait i didn't say i preferred single underscore. tbh i don't know which i prefer was hoping for u to choose one lol.

File: backend/portal/serializers.py
Comment (vcai122) [reply backend/portal/serializers.py:32]: Good thought, although it doesn't matter here since if you look Kind has like 4 values. Probably the overhead from sets is more costly anyway

File: backend/portal/models.py
Comment (tuneerroy) [reply backend/portal/models.py:58]: lol i'd just stick with SO's opinion, single underscore for private methods, double underscore if you're concerned about inheritance overriding

**Pull request merged**
